### PR TITLE
feat(framework): add nuxt framework client

### PR DESCRIPTION
**Pull request by peterbud**
## ☕️ Reasoning

This PR is aiming to add support for Nuxt framework to `@AuthJS` 

## 🧢 Checklist

- [x] Add `SetEnvDefaults` to be consistent with other frameworks
- [x] Discuss CSRF vs Origin checking approach
- [X] Documentation
- [x] Ready to be merged

## 🎫 Affected issues

N/A

### CODE REVIEW

**Comment by ndom91** [comment]
Followup Todos for us:
- [ ] Add example repo (`nextauthjs/nuxt-example-app`)
- [ ] Ensure Balazs github token can write to that (for sync)
- [ ] Double check example code sync to that new repo
- [ ] Add nuxt example app Vercel project (with shared env vars)

File: package.json
Comment (ndom91) [inline package.json:61]: Do `eslint-plugin-vue` and `vue-eslint-parser` need to be int he root `package.json`? Ideally these would be in the `framework-nuxt/package.json` instead

Comment (peterbud) [reply package.json]: OK, good point. I have moved them.

Comment (peterbud) [reply package.json]: On a second note: we would need this also in the apps/dev/nuxt project as well, right?

**Comment by peterbud** [comment]
> is it possible to build it and use it as alpha until the production version released ? @peterbud

I think we are pretty early in the process of creating this framework, I'd love to have a proper review from the AuthJS core members before suggesting to use this.

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (UfukUstali) [inline packages/frameworks-nuxt/src/runtime/lib/server.ts:74]: Shouldn't this be `${config.basePath}/session`

I can see that having to pass in config as an argument everytime one of the utility functions is called could be annoying

My suggestion would be to change the signature to be 
```ts
async function getServerSession(event: H3Event, config: authConfig)
// ^ not exported
```

and do the same to any other utilities that will be written

after that using the function below

```ts
export function generateServerUtils(config: authConfig) {
  return {
    getServerSession: async (event: H3Event) => await getServerSession(event, config),
  }
}
```

the consumer can generate their own utilities in `~/server/utils/` folder which will then be auto imported by nuxt without any further assistance.
see -> [auto-imports](https://nuxt.com/docs/guide/concepts/auto-imports) and [server-utilities](https://nuxt.com/docs/guide/directory-structure/server#server-utilities)

```ts
// ~/server/utils/auth.ts
import { generateServerUtils } from "#auth"
import { authConfig } from "@/server/[api/auth/, routes/auth/, routes/whatever][...].ts"

export const { getServerSession } = generateServerUtils(authConfig );

```

Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/lib/server.ts:74]: Thanks for the review, I'll get to it over the weekend

**Comment by vinayakkulkarni** [comment]
Oh man! This merge would be epic! I could do a major refactor :D

File: apps/dev/nuxt/.eslintrc.cjs
Comment (vinayakkulkarni) [inline apps/dev/nuxt/.eslintrc.cjs:25]: Maybe integrate the official ESLint module – https://eslint.nuxt.com/packages/module ?

Comment (vinayakkulkarni) [reply apps/dev/nuxt/.eslintrc.cjs:25]: Blog article by the great @antfu – https://nuxt.com/blog/eslint-module

Comment (peterbud) [reply apps/dev/nuxt/.eslintrc.cjs:25]: Eslint rules and config has been defined for the full monorepo, I don't think the scope of this change would be to change that.

**Comment by shaharyarrahmad** [comment]
Any Update on this PR?

**Comment by peterbud** [comment]
> Any Update on this PR?

Still waiting for the review of the core team.

File: packages/frameworks-nuxt/src/runtime/middleware/auth-logged-out.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/middleware/auth-logged-out.ts:1]: Can we remove this file? This looks identical to `auth-logged-in` - IMO we should rename the middleware to just `auth`

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/lib/server.ts:66]: Let's rename this to `auth` to match with other framework integrations.

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/lib/server.ts:24]: How do you think if we add a combined export like this:
```ts
export function NuxtAuth(config: AuthConfig) {
  return {
    handlers: // the current code for `NuxtAuthHandler`
    auth: // the current code for `auth` aka `getServerSession`
  }
}
```

Later on in the `server/util/auth` file:
```ts
const { handlers, auth as _auth } = NuxtAuth(config)

export function authHandler() {
  return handlers
}

export async function auth(event: H3Event) {
  return await _auth(event)
}


```
It will look more unified with the NextAuth init:
Screenshot 2024-10-27 at 15 45 24

File: packages/frameworks-nuxt/src/module.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/module.ts:12]: will there be cases where we need to customize this value.

File: packages/frameworks-nuxt/src/runtime/plugin.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/plugin.ts:21]: shouldn't it be
```suggestion
      else removeSession()
```

File: packages/frameworks-nuxt/src/runtime/utils.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/utils.ts:21]: unused function

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (ThangHuuVu) [inline packages/frameworks-nuxt/src/runtime/lib/server.ts:44]: why don't we just set it as: 
```suggestion
    config.basePath ??= '/auth'
```
like other frameworks.

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]

**Comment by moshetanzer** [comment]
Yay like to see review! We need this already! Thanks!

File: packages/frameworks-nuxt/src/runtime/middleware/auth-logged-out.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/middleware/auth-logged-out.ts:1]: These middlewares are actually the _opposite_ of each other. These are _optional_ middlewares, which are available for the users of the module, in case they don't want to create their own middleware. In that case users can use middlewares defined in the module:
* If `auth-logged-in` middleware is configured for a certain route it would require an authenticated, logged in user for the given route, otherwise navigation would be aborted.
* If `auth-logged-out` middleware is configured for a certain route it would work only unauthenticated users, aka authenticated users cannot access those pages/routes

I find these useful additions to the modules and they are similar to other nuxt auth modules (like https://github.dev/nuxt-modules/kinde)

File: packages/frameworks-nuxt/src/runtime/plugin.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/plugin.ts:21]: Good catch!

File: packages/frameworks-nuxt/src/runtime/utils.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/utils.ts:21]: Removed

File: packages/frameworks-nuxt/src/module.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/module.ts:12]: Not really. Would you like to change it to a simple const?

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/lib/server.ts:24]: I think that could work. Let me refactor

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/lib/server.ts:44]: OK

File: packages/frameworks-nuxt/src/runtime/lib/server.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/runtime/lib/server.ts]: Done

**Comment by peterbud** [comment]
@ThangHuuVu , I have resolved all the conversations which I was able to resolve, and responded to the others. Let me know how to proceed

File: packages/frameworks-nuxt/src/module.ts
Comment (peterbud) [reply packages/frameworks-nuxt/src/module.ts]: I modified to use simple 'auth' as configKey

**Comment by ndom91** [comment]
Follow up for us once this is merged:

- [ ] Add `nuxt.authjs.dev` subdomain in Vercel dashboard (redirect already in this PR)
- [ ] Add example app 
	- [ ] Create repository `nextauthjs/nuxt-auth-example`
	- [ ] Add new repo to `.github/sync.yml` 
	- [ ] Create new Vercel project syncing from new `nextauthjs/nuxt-auth-example` repo
	- [ ] Link pre-existing account env vars (esp shared `AUTH_SECRET` for proxying)
	- [ ] Ensure the `apps/examples/nuxt` dir is syncing to `nextauthjs/nuxt-auth-example` correctly (Balazs often has to add the new repo to his PAT for the sync to work, otherwise permissions errors)
	- [ ] Add example app link to docs homepage (?) 
- [ ] Tweet about it :partying_face: (?) 

Anything I've missed?

**Comment by moshetanzer** [comment]
@ndom91 
Not sure but removing NUXT_ from beginning of env vars will not let them be replaced in runtime config properly.

**Comment by peterbud** [comment]
> @ndom91 Removing NUXT_ from beginning of env vars will not let them be replaced in nuxt config in runtime properly.

Yes, without the `NUXT_` prefix they won't be read from the .env file and at runtime they won't be even considered.

**Comment by ndom91** [comment]
Okay gotcha, I'll revert that then. Unfortunately the auto-detect env vars from `@auth/core` wont work then without additional changes there, but that's okay for now imo

**Review by ndom91 — APPROVED** [review]

**Comment by orenmizr** [comment]
i noticed next-auth changed in v5. will this be v5 compatible ? will this be offical nuxt or community ? 

kudos @peterbud for sticking with it all this time.

**Comment by peterbud** [comment]
> i noticed next-auth changed in v5. will this be v5 compatible ? 

v5 compatible

**Comment by peterbud** [comment]
I'm not sure what else would be needed to move forward with this PR.

It has been open for more than 6 months, and recently I looked optimistic as we finally managed to get the reviews completed with the help of the core team.

In the last 2 weeks this PR seems to be stalled again, and it's again drifting from the main branch.

Could you be please answer the question honestly and make a decision on whether there is more effort to be done into this PR (and what that would be), or make the decision to cancel this PR. At this point I think canceling this PR is probably more respectful to each others time and energy.

**Comment by gsxdsm** [comment]
Would love to see this merged in

**Comment by gsxdsm** [comment]
@ThangHuuVu can this please be merged? Thanks!

**Comment by gsxdsm** [comment]
Agree @peterbud thank you so much!

**Comment by peterbud** [comment]
> not sure why "authjs" has been ignoring nuxt. it is the closest the next. in this time, a lot of devs (myself included) turned to nuxt-auth-utils as the alternative

I agree. I think I have to let this go.

**Comment by moshetanzer** [comment]
@peterbud really appreciate all the effort. Really such a disgrace of the team!

**Comment by gsxdsm** [comment]
Agree thanks @peterbud ! Based on how this was handled I moved to better-auth. Works superbly.

**Comment by zoey-kaiser** [comment]
Hi @peterbud 👋

I’ve been following the progress of this PR and was really excited to see where it was heading. You’ve done an amazing job with the integration, and it’s honestly a bit sad to see the PR closed.

I work on a Nuxt 3 module for NextAuth called [@sidebase/nuxt-auth](https://github.com/sidebase/nuxt-auth), and we’ve tried several times to update it to work with AuthJS. Unfortunately, we've run into some stability issues along the way.

Seeing how much progress you’ve made, I wanted to reach out and see if you'd be interested in helping us tackle some of the issues we’ve faced with the upgrade. It would be awesome to release a Nuxt 3 integration of AuthJS for the community!

If you're interested, feel free to DM me on BlueSky: https://bsky.app/profile/zoeykaiser.com

**Comment by peterbud** [comment]
Hi @zoey-kaiser,
 
Thank you so much for your kind words and for following the progress of the PR. I truly appreciate your support and enthusiasm.

I understand the challenges you've faced with updating your Nuxt 3 module to work with AuthJS. However, at this time, I must politely decline your request for assistance. The future of AuthJS seems uncertain, and the maintainers appear to be somewhat disconnected, which makes it difficult to commit to this integration.

I wish you the best of luck with your project and hope you find a suitable solution soon.

**Comment by AdnanCukur** [comment]
This is critical for our use case. We’d love to see this merged please let us know if we can help with testing or resolving any blockers!

**Comment by moshetanzer** [comment]
@AdnanCukur don't waste your time. The team were rude and unhelpful tk @peterbud who put in the effort. This has fallen way behind. Rather use something like better auth

**Comment by makkalay** [comment]
that work nice with some tweaks, https://authjs-nuxt.pages.dev/getting-started

**Pull request closed without merging**
